File: GermanyWeb/src/js/deck.js
const SUITS = ["♠","♥","♦","♣"];
const VALUES = "A,2,3,4,5,6,7,8,9,10,J,Q,K".split(",");

export default class Deck {
    constructor(cards = generateDeck()){
        this.cards = cards;
    }
    emptyDeck(){
        this.cards.splice(0,this.cards.length)
    }
    shuffle(){
        for (let i = this.size-1; i>0; i--){
            //get a random card to swap with
            const target = Math.floor(Math.random() * (i+1));
            //swap with the random card
            const oldValue = this.cards[target];//save the random card's value
            this.cards[target] = this.cards[i];//set the random card's value to this card's value
            this.cards[i] = oldValue;//set this card's value to the saved random card value
        }
    }
    //size of the deck tends to be referenced a lot, so this is a getter for that
    get size(){
        return this.cards.length;
    }
    get topCard(){
        return this.cards[0]
    }
    collectMultiple(cards){
        this.cards = [...this.cards,...cards]
    }
    collectSingle(card){
        this.cards.unshift(card)
    }
    remove(card){
        const index = this.cards.indexOf(card)
        if (index > -1){
            this.cards.splice(index,1)
        }
    }
}

class Card {
    constructor(suit, value){
        this.suit = suit;
        this.value = value;
    }
   
    get color(){
        if (this.value === "7" || this.value === "10" || this.value === "2") {
            return "golden";
        } else if (this.suit==="♠" || this.suit==="♣"){
            return "black";
        } else{
            return "red";
        } 
    }
    getHTML(){//return an card as an html element with this format: <div class="card ${color}" data-value="${value} ${suit}">${suit}</div>
        const cardElement = document.createElement('div');
        cardElement.innerText = this.suit;
        cardElement.classList.add('card',this.color);
        cardElement.dataset.value = `${this.value} ${this.suit}`;
        return cardElement;
    }
    get toString(){
        return this.value + " " + this.suit 
    }
}
function generateDeck(){
    return SUITS.flatMap(suit => {
        return VALUES.map(value => {
            return new Card(suit, value);
        });
    });
}

File: Breakout/deck/deck.go
package deck

import (
	"math/rand"
	"strings"
	"time"
)

type Deck struct {
	Cards []*Card `json: "cards"`
}

func NewDeck() *Deck {
	suits, values := [4]string{"♠", "♥", "♦", "♣"}, strings.Split("2,3,4,5,6,7,8,9,10,J,Q,K,A", ",")
	deck := []*Card{}
	for _, suit := range suits {
		for n, value := range values {
			deck = append(deck, NewCard(suit, value, n+1))
		}
	}
	return &Deck{deck}
}
func EmptyDeck() *Deck {
	return &Deck{[]*Card{}}
}

func (d *Deck) Size() int {
	return len(d.Cards)
}
func (d *Deck) Top() *Card {
	return d.Cards[0]
}
func (d *Deck) Collect(cards ...*Card) {
	if len(cards) == 1 {
		d.Cards = append(cards, d.Cards...)
	} else {
		d.Cards = append(d.Cards, cards...)
	}
}
func (d *Deck) Shuffle() {
	//new seed every time Shuffle() is called to prevent the same deck from being used over and over agagin
	rand.Seed(time.Now().UnixNano())
	for i := d.Size() - 1; i > 0; i-- {
		target := rand.Intn(i)
		oldValue := d.Cards[target]
		d.Cards[target] = d.Cards[i]
		d.Cards[i] = oldValue
	}
}
func (d *Deck) Remove(c *Card) *Card {
	//find the index
	i := d.indexOf(c)
	if i >= 0 {
		rv := d.Cards[i]
		//splice it out
		d.Cards = append(d.Cards[:i], d.Cards[i+1:]...)
		return rv
	} else {
		return nil
	}
}

func (d *Deck) indexOf(c *Card) int {
	for i, card := range d.Cards {
		if card.Suit == c.Suit && c.Value == card.Value {
			return i
		}
	}
	return -1
}
